runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -2180,26 +2180,33 @@ J9::TransformUtil::refineMethodHandleLinkTo(TR::Compilation* comp, TR::TreeTop*
    TR::ILOpCodes callOpCode = getTargetMethodCallOpCode(rm, node->getDataType());
 
    TR::SymbolReference* newSymRef = NULL;
+   uint32_t vTableSlot = 0;
+   int32_t jitVTableOffset = 0;
    if (rm == TR::java_lang_invoke_MethodHandle_linkToVirtual)
       {
-      uint32_t vTableSlot = fej9->vTableOrITableIndexFromMemberName(comp, mnIndex);
-      auto resolvedMethod = fej9->createResolvedMethodWithVTableSlot(comp->trMemory(), vTableSlot, targetMethod, symRef->getOwningMethod(comp));
-      newSymRef = comp->getSymRefTab()->findOrCreateMethodSymbol(symRef->getOwningMethodIndex(), -1, resolvedMethod, callKind);
-      newSymRef->setOffset(fej9->vTableSlotToVirtualCallOffset(vTableSlot));
-      }
-   else
-      {
-      uint32_t vTableSlot = 0;
-      auto resolvedMethod = fej9->createResolvedMethodWithVTableSlot(comp->trMemory(), vTableSlot, targetMethod, symRef->getOwningMethod(comp));
-      newSymRef = comp->getSymRefTab()->findOrCreateMethodSymbol(symRef->getOwningMethodIndex(), -1, resolvedMethod, callKind);
-      }
+      vTableSlot = fej9->vTableOrITableIndexFromMemberName(comp, mnIndex);
+      jitVTableOffset = fej9->vTableSlotToVirtualCallOffset(vTableSlot);
+      // For private virtual methods and java/lang/Object; type virtual methods, there is no corresponding
+      // entry in the vtable, and for such methods the interpreter vtable index is 0.
+      // The dispatch is not performed through the vtable entry, but directly dispatched to the J9Method
+      // in Membername.vmtarget. In such cases, we can refine a linkToVirtual call similar to linkToStatic.
+      // jitVTableOffset is obtained using sizeof(J9Class) - interpreter vtable offset, resulting in a positive
+      // value when the interpreter vtable index we get (from MemberName.vmindex.vtableoffset) is set as 0.
+      if (jitVTableOffset > 0)
```

Comment:
I'm confused by this test.  If the problematic interpreter vtable offset is "sizeof(J9Class)" or +376 then shouldn't this test be `jitVTableOffset >= 0` ?

Reply:
The problematic interpreter vtable offset is when it is set as `0`, resulting in the jitVtableOffset being `sizeof(J9Class)`. Normally, interpreter vtable offset is set a value `> sizeof(J9Class)` so the vtable offset we obtain will always be negative, which is why I thought a simple test for a positive integer value was sufficient. We can't have `jitVtableOffset == 0` as that is the start of the J9Class. Do you think changing the operator to `>=` would provide more clarity?

Reply:
Just for reference, here is where the interpreter vtable offset is set:
https://github.com/eclipse-openj9/openj9/blob/7ad86703975e3432b6ec90b998ab04aa1a835e71/runtime/vm/createramclass.cpp#L1407-L1421

`J9VTABLE_OFFSET_FROM_INDEX` takes the index in the vtable and adds `sizeof(J9Class) + sizeof(J9VTableHeader)`, and the vtable entries begin following the J9VTableHeader. So whenever there's a vtable entry, the jitVTableOffset will end up being a negative value.